[AArch64][PAC] Lower authenticated calls with ptrauth bundles.

ahmedbougacha opened the pull request:
This adds codegen support for the "ptrauth" operand bundles, which can
be used to augment indirect calls with the equivalent of an
`@llvm.ptrauth.auth` intrinsic call on the call target (possibly
preceded by an `@llvm.ptrauth.blend` on the auth discriminator if
applicable.)

This allows the generation of combined authenticating calls
on AArch64 (in the BLRA* PAuth instructions), while avoiding
the raw just-authenticated function pointer from being
exposed to attackers.

This is done by threading a PtrAuthInfo descriptor through
the call lowering infrastructure.

Note that this also applies to the other forms of indirect calls,
notably invokes, rvmarker, and tail calls.  Tail-calls in particular
bring some additional complexity, with the intersecting register
constraints of BTI and PAC discriminator computation.

This also adopts an x8+ allocation order for GPR64noip, matching GPR64.

ahmedbougacha committed:
[AArch64] Adopt x8+ allocation order for GPR64noip.

73078ecd381 added GPR64noip for hwasan pseudos.
Give it an allocation order that prefers allocating from x8 and up,
to match GPR64: this allows for easier regalloc, as x0-x7 are
likely to be used for parameter passing.

ahmedbougacha committed:
[AArch64][PAC] Lower authenticated calls with ptrauth bundles.

This adds codegen support for the "ptrauth" operand bundles, which can
be used to augment indirect calls with the equivalent of an
`@llvm.ptrauth.auth` intrinsic call on the call target (possibly
preceded by an `@llvm.ptrauth.blend` on the auth discriminator if
applicable.)

This allows the generation of combined authenticating calls
on AArch64 (in the BLRA* PAuth instructions), while avoiding
the raw just-authenticated function pointer from being
exposed to attackers.

This is done by threading a PtrAuthInfo descriptor through
the call lowering infrastructure.

Note that this also applies to the other forms of indirect calls,
notably invokes, rvmarker, and tail calls.  Tail-calls in particular
bring some additional complexity, with the intersecting register
constraints of BTI and PAC discriminator computation.

ahmedbougacha commented:
ping

kovdan01 commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -1909,9 +1948,60 @@ void AArch64AsmPrinter::emitInstruction(const MachineInstr *MI) {
+  case AArch64::AUTH_TCRETURN:
+  case AArch64::AUTH_TCRETURN_BTI: {
+    const uint64_t Key = MI->getOperand(2).getImm();
+    assert(Key < 2 && "Unknown key kind for authenticating tail-call return");
```
There are many assertions and if statements all around the code which try to check that key is either IA or IB. They look inconsistent: sometimes we have `Key < 2`, sometimes `Key == 0 || Key == 1', sometimes `Key > 1` (for an opposite check). I like the approach from `AArch64AsmPrinter::emitPtrauthBranch`:

```
  assert((Key == AArch64PACKey::IA || Key == AArch64PACKey::IB) &&
         "Invalid auth call key");`
```

It does not rely on magic constants and it's our intentions are clear. The check might be moved to a small helper, it'll be probably more handy to use it everywhere. Please make assertions and if statements checking the key against IA or IB consistent with each other.

ahmedbougacha committed:
Address review feedback.

- test invoke for ELF as well
- consolidate GISel/SDAG invoke test
- check key/disc validity in consistent ways
- misc. nits

ahmedbougacha commented on llvm/lib/Target/AArch64/AArch64InstrInfo.td:
```diff
@@ -1694,6 +1718,35 @@ let Predicates = [HasPAuth] in {
 
+  // BLRA pseudo, generalized version of BLRAA/BLRAB/Z.
+  // This directly manipulates x16/x17, which are the only registers the OS
+  // guarantees are safe to use for sensitive operations.
```
Unfortunately this sort of thing isn't documented.  We do describe the general contract in the security model sections of the clang docs.
For ELF it might make sense to have a formalization in the PAuth ABI, but that would have to be actually honored in an OS;  I imagine it currently isn't anywhere other than Darwin.

kovdan01 replied:
Yes, having a more detailed comment here would be nice. As for now, from the comment it was unclear that this is Darwin-specific. It's also probably worth mentioning that it's harmless for non-Darwin, as you said.

ahmedbougacha commented on llvm/test/CodeGen/AArch64/GlobalISel/ptrauth-invoke.ll:
```diff
@@ -0,0 +1,183 @@
+; RUN: llc -mtriple arm64e-apple-darwin -o - -global-isel -global-isel-abort=1 -verify-machineinstrs %s | FileCheck %s --check-prefixes=CHECK
```
It's not standard practice to test other platforms, but sure, I added ELF checks.  Now you folks owe me darwin tests for future PAC changes you make ;)

ahmedbougacha commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8640,6 +8642,15 @@ void SelectionDAGBuilder::LowerCallTo(const CallBase &CB, SDValue Callee,
+  // Set the pointer authentication info if we have it.
+  if (PAI) {
+    if (!TLI.supportPtrAuthBundles())
+      report_fatal_error(
```
I don't think testing this one is reasonable;  we couldn't even put it in AArch64 and would have to pollute another innocent backend

ahmedbougacha commented:
Thanks for taking a look;  updated

kovdan01 commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -817,10 +817,44 @@ bool AArch64ExpandPseudo::expandCALL_RVMARKER(
+    // Skip register arguments. Those are added during ISel, but are not
+    // needed for the concrete branch.
+    while (!MI.getOperand(RegMaskStartIdx).isRegMask()) {
+      auto MOP = MI.getOperand(RegMaskStartIdx);
```
Just above `auto` is not used in the same kind of assignments. IMHO things should be kept consistent, so either `const auto &` or `const MachineOperand &` should be used for all such cases in this context.

ahmedbougacha replied:
The helper indeed shouldn't use plain auto, but we can reuse that helper with a bit of generalization.  I'm pretty sure this code predates it ;)

kovdan01 replied:
Please see overall comment

kovdan01 commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -817,10 +817,44 @@ bool AArch64ExpandPseudo::expandCALL_RVMARKER(
+  if (MI.getOpcode() == AArch64::BLRA_RVMARKER) {
+    // Pointer auth call.
+    MachineOperand &Key = MI.getOperand(2);
+    assert((Key.getImm() == 0 || Key.getImm() == 1) &&
```
Please use `AArch64PACKey::I{A|B}` instead of 0 and 1

kovdan01 commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -817,10 +817,44 @@ bool AArch64ExpandPseudo::expandCALL_RVMARKER(
+
+  if (MI.getOpcode() == AArch64::BLRA_RVMARKER) {
+    // Pointer auth call.
+    MachineOperand &Key = MI.getOperand(2);
```
```suggestion
    const MachineOperand &Key = MI.getOperand(2);
```

kovdan01 replied:
Please see overall comment

kovdan01 commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -817,10 +817,44 @@ bool AArch64ExpandPseudo::expandCALL_RVMARKER(
+    MachineOperand &Key = MI.getOperand(2);
+    assert((Key.getImm() == 0 || Key.getImm() == 1) &&
+           "invalid key for ptrauth call");
+    MachineOperand &IntDisc = MI.getOperand(3);
```
```suggestion
    const MachineOperand &IntDisc = MI.getOperand(3);
```

kovdan01 replied:
Please see overall comment

kovdan01 commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -817,10 +817,44 @@ bool AArch64ExpandPseudo::expandCALL_RVMARKER(
+    assert((Key.getImm() == 0 || Key.getImm() == 1) &&
+           "invalid key for ptrauth call");
+    MachineOperand &IntDisc = MI.getOperand(3);
+    MachineOperand &AddrDisc = MI.getOperand(4);
```
```suggestion
    const MachineOperand &AddrDisc = MI.getOperand(4);
```

kovdan01 replied:
Please see overall comment

kovdan01 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -329,6 +329,40 @@ static bool isZeroingInactiveLanes(SDValue Op) {
+  // If the constant discriminator (either the blend RHS, or the entire
+  // discriminator value) isn't a 16-bit constant, bail out, and let the
+  // discriminator be computed separately.
+  auto *ConstDiscN = dyn_cast<ConstantSDNode>(ConstDisc);
```
```suggestion
  const auto *ConstDiscN = dyn_cast<ConstantSDNode>(ConstDisc);
```

kovdan01 reviewed (dismissed):
@ahmedbougacha In terms of functionality, LGTM (but I'll prefer other guys also looking through the changes). I've left several minor style-related comments.

One more overall style comment which might probably require an additional PR: sometimes identifiers use `PtrAuth`, sometimes `Ptrauth`. Is this intentional? If not, I suggest to use the same case style - it would keep things consistent and allow for case-sensitive search.

ahmedbougacha committed:
Address review feedback.

- AArch64ExpandPseudos: generalize & use createCall
- AArch64InstrInfo: describe x16/x17 usage in pseudo
- AArch64InstrInfo: group SDNodes with others
- test call for ELF as well
- GlobalISel: rename PointerAuthInfo, reorder fields
- various nits, auto, const

ahmedbougacha commented:
> One more overall style comment which might probably require an additional PR: sometimes identifiers use `PtrAuth`, sometimes `Ptrauth`. Is this intentional? If not, I suggest to use the same case style - it would keep things consistent and allow for case-sensitive search.

Yeah, it's hard to avoid needing case-insensitive search because of the irreconcilable capitalization rules between e.g., enums generated from all-lowercase tablegen, opcode namespaces with a strong all-uppercase tradition, variable/function naming conventions, etc.

In general I try to standardize on "ptrauth" regardless of case, and would rather get rid of "pauth" or "pac" first, because they're just objectively worse names.  We can indeed pick a consistent ptrauth when the dust settles.

ahmedbougacha commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -1769,6 +1775,41 @@ void AArch64AsmPrinter::emitPtrauthAuthResign(const MachineInstr *MI) {
 
+void AArch64AsmPrinter::emitPtrauthBranch(const MachineInstr *MI) {
+  unsigned InstsEmitted = 0;
+
```
It's to try to isolate the InstsEmitted prologue/epilogue from the real logic.  Ideally that could be some sort of RAII helper, which we could probably use in all the expanders, but for now, ad-hoc it is, without the newline I suppose

kovdan01 reviewed (commented):
LGTM with a few const-related comments left (actually I've unresolved several previously opened threads). I highly encourage everyone else interested to also look through since while the changes look reasonable and the code does what it's intended to do, there might be some corner cases I've missed or other room for enhancement.

@ahmedbougacha Do I miss some reason why const qualifier should not be applied to pointers/references I've mentioned? Even though it might be obvious for experienced llvm contributors that these values are not modified, it's still IMHO better to explicitly mark them as const so a person not so experienced in this area would not have to guess while reading the code whether the non-const-qualified value is intended to be modified when passed somewhere by pointer/reference or not.

ahmedbougacha committed:
Minor refinements.

- assert ptrauth calls aren't used with PAuth_LR
- more const nits

ahmedbougacha commented:
I rebased this on top of main and extracted the couple of common helper functions from the auth/resign isel patch, to not depend on it.

> LGTM with a few const-related comments left

Thanks, all of these should be addressed now.  I'll leave this open for another day or two if anyone wants to chime in, either way we can iterate further in tree
